### PR TITLE
Core: Enhance weightFunc of bin-packing to adapt to V2Format

### PR DESCRIPTION
**Pull request by WinkerDu**
In our scene, We use V2Format to support streaming CDC row-level insert / delete. There could be tons of delete file for scan task of small file rewrite action or query scan, such as Spark Batch scan typically.

The existing scan task bin-packing logic is only based on data file size, means a scan task contains data files whose total file size satisfies a given target size, this logic work fine for V1Format since the task only deals with data file.

But for V2Format, scan task performance also consists of delete file applying cost. Suppose that bin-packing target size is 128MB, the data file could be small due to streaming CDC update, such as 1 MB, each data file tries to apply 128 valid eq-delete / pos-delete files. the total delete file applied number of 1 task could reach to 128 * 128 = 16,384, even we have enough CPU core to run these tasks, we could not boost scan performance for 1 task.

This PR enhance weight function to take into account how much of a penalty a delete file is, and solve unbalanced bin-packing issue.

### CODE REVIEW

**Comment by WinkerDu** [comment]
@rdblue @openinx could you please review this patch?

**Comment by RussellSpitzer** [comment]
I think this is a good feature to add but perhaps the name should be something along the lines of 
`items-per-bin` or `max-items-per-bin` ? 

Or maybe we could hit this use case directly and set a max number of `delete files` referenced per bin?

**Comment by WinkerDu** [comment]
@RussellSpitzer thanks for your replay. The name has changed to `items-per-bin`

I think the number of data files and delete referenced per bin should be specified separately, `insert` operation only add data files for example. I am glad to deal with delete-files situation in following PR

**Comment by WinkerDu** [comment]
@aokolnychyi @kbendick @jackye1995 please have a further review

**Comment by RussellSpitzer** [comment]
I'll take a full pass on tuesday, I'm just getting settled back home but we have power and interenet so should be plenty of time then.

**Comment by WinkerDu** [comment]
image
checked the ci log, not sure why it was cancelled. Can I retry this ci? @RussellSpitzer

**Comment by RussellSpitzer** [comment]
I think this is the timeout issue me and @kbendick were discussing yesterday

**Comment by RussellSpitzer** [comment]
I wouldn't worry about doing a retry till we get the reviews done, if it passed java 11 (and it did) it just means it is an unrelated issue I think

**Comment by WinkerDu** [comment]
@RussellSpitzer OK, let's get the review done first :)

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 1]
```diff
@@ -65,6 +65,7 @@
   private Expression filter;
   private long targetSizeInBytes;
   private int splitLookback;
+  private int itemsPerBin;
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:68]: Looks like this is only modifying the old rewrite path and not the new path. We should probably fix it in both places

File: core/src/main/java/org/apache/iceberg/util/BinPacking.java [hunk 2]
```diff
@@ -169,19 +179,21 @@ public boolean hasNext() {
...
+      this.itemsPerBin = itemsPerBin;
     }
 
     List<T> items() {
       return items;
     }
 
     boolean canAdd(long weight) {
-      return binWeight + weight <= targetWeight;
+      return binWeight + weight <= targetWeight && items.size() <= itemsPerBin;
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/util/BinPacking.java:196]: I think there is a slight issue with this implementation and how it interacts with lookback.

Consider we have a lookback of 3 a max number of items per bin of 2 and a max value of bin of 10
```
// Add Item (6)
[ 8 ], [3, 4], [4, 3]
// Even if remove largest is set largest bin is [8] but we actually have 2 bins we know we can never add another item to which we should be dropping first.

I think to properly do this we need to also finish off any bins which have reached their max item size
```

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java (hunk 1 above)
Comment (WinkerDu) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:68]: Well, what new path you mean, `RewriteDataFiles` and related implements?

File: core/src/main/java/org/apache/iceberg/util/BinPacking.java (hunk 2 above)
Comment (WinkerDu) [reply core/src/main/java/org/apache/iceberg/util/BinPacking.java:196]: OK, I will push a new commit to do this with UT case

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java (hunk 1 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:68]: Yeah the new api extends off the interface rather than this base class.

**Comment by openinx** [comment]
Let's retry the travis CI ..

**Comment by openinx** [comment]
@WinkerDu I definitely agreed the v2 bin-pack algorithm should be improved for v2 to consider the total size of insert & delete files.  I think the `iterms-per-bin` proposed from you team is trying to resolve the unbalanced issue,  but I'm concerning it's hard to set the correct `iterms-per-bin` value for a given table in real production environment,  because the `iterms-per-bin` is still controlling the data file's count.  We actually don't have a real suitable approach to evaluate the cost about joining the data file size & its delete records.  I think we need more accurate approach to decide which scan tasks should be dispatched to different tasks.

**Comment by RussellSpitzer** [comment]
I agree with @openinx that ideally we really should try to get a more V2-style algorithm here that is able to take into account deletes and such. But I don't have an issue with adding more parameters to bin-pack.

File: core/src/main/java/org/apache/iceberg/util/BinPacking.java [hunk 3]
```diff
@@ -36,51 +36,59 @@
...
     private final boolean largestBinFirst;
 
     public ListPacker(long targetWeight, int lookback, boolean largestBinFirst) {
+      this(targetWeight, lookback, largestBinFirst, Integer.MAX_VALUE);
+    }
+
+    public ListPacker(long targetWeight, int lookback, boolean largestBinFirst, int itemsPerBin) {
       this.targetWeight = targetWeight;
       this.lookback = lookback;
+      this.itemsPerBin = itemsPerBin;
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/util/BinPacking.java:49]: Nit: might want to move this one below `largestBinFirst` so they're in the same order as the parameters list.

**Review by kbendick — COMMENTED** [review]
I need to set aside some time to take a closer look at some of these tests. Will do so tonight.

**Comment by kbendick** [comment]
> @WinkerDu I definitely agreed the v2 bin-pack algorithm should be improved for v2 to consider the total size of insert & delete files. I think the `iterms-per-bin` proposed from you team is trying to resolve the unbalanced issue, but I'm concerning it's hard to set the correct `iterms-per-bin` value for a given table in real production environment, because the `iterms-per-bin` is still controlling the data file's count. We actually don't have a real suitable approach to evaluate the cost about joining the data file size & its delete records. I think we need more accurate approach to decide which scan tasks should be dispatched to different tasks.

I need to spend some time looking closer at the test cases (and probably try this out on some V2 tables), but I share the concern that this config value might be really hard to determine in a production env. Especially for example Flink users who have CDC streams, often times databases will experience a burst of deletes / updates due to some cron schedule and will then have an outsized number of delete files for a period of time (assuming partitioned by time as well).

Wondering how we would go about picking a good number (or how often one would need to set a non-standard number, or change the number for individual sections of the table).

That said, I'm also not adverse to adding another argument to make bin packing more useful in the near-term while we figure out the best way to have a more "V2 native" algorithm / parameter set.

**Comment by rdblue** [comment]
This doesn't seem like the right way to solve the problem that some splits are too expensive because there are too many delete files. A symptom of the problem is that it isn't obvious how to set the maximum number of items per bin.

I think that the right approach to solve the problem is to modify the weight function that we use for splits, _not_ to modify the bin packing algorithm itself.

A straightforward way to fix this problem is to make the weight function take into account how much of a penalty a delete file is. Right now, the weight function is this, in `TableScanUtil`:

```java
    Function<FileScanTask, Long> weightFunc = file -> Math.max(file.length(), openFileCost);
```

That could easily be updated so that each delete file that must be opened is added to the open cost:

```java
    Function<FileScanTask, Long> weightFunc = file -> Math.max(file.length(), (1 + file.deletes().size()) * openFileCost);
```

Another option is to check the size of the delete data as well:

```java
    Function<FileScanTask, Long> weightFunc = file -> Math.max(
        file.length() + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
        (1 + file.deletes().size()) * openFileCost);
```

Let's try those options instead of modifying bin packing. I think modifying the packing algorithm is the wrong direction.

**Comment by WinkerDu** [comment]
@openinx @kbendick @rdblue Thanks for your reply.
I think `weightFunc` @rdblue suggests is a good option to solve the problem you mentioned.
Will push a new commit later.

**Comment by WinkerDu** [comment]
@rdblue @RussellSpitzer @openinx @kbendick A new commit has been pushed, please have a review.

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 4]
```diff
@@ -211,7 +220,13 @@ protected long writeMaxFileSize() {
   }
 
   private long sizeOfInputFiles(List<FileScanTask> group) {
-    return group.stream().mapToLong(FileScanTask::length).sum();
+    return group.stream().mapToLong(this::sizeOfInputFile).sum();
+  }
+
+  private long sizeOfInputFile(FileScanTask file) {
+    // For V2Format, we should check the size of delete file as well to avoid unbalanced bin-packing
+    return Math.max(file.length() + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:228]: Can you wrap this so that both arguments to `max` start on a new line?

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 5]
```diff
@@ -234,5 +249,8 @@ private void validateOptions() {
     Preconditions.checkArgument(minInputFiles > 0,
         "Cannot set %s is less than 1. All values less than 1 have the same effect as 1. %d < 1",
         MIN_INPUT_FILES, minInputFiles);
+
+    Preconditions.checkArgument(openFileCost > 0,
+        "Cannot set %s is less than 1, %d < 1", RewriteDataFiles.OPEN_FILE_COST, openFileCost);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:254]: This should be "Invalid file open cost: %d (not positive)"

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 6]
```diff
@@ -211,7 +220,13 @@ protected long writeMaxFileSize() {
   }
 
   private long sizeOfInputFiles(List<FileScanTask> group) {
-    return group.stream().mapToLong(FileScanTask::length).sum();
+    return group.stream().mapToLong(this::sizeOfInputFile).sum();
+  }
+
+  private long sizeOfInputFile(FileScanTask file) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:226]: @RussellSpitzer, what do you think about this change to `BinPackStrategy`?

I'm not sure that we want to make this change to the file rewrite. While delete files are relevant to how this will run, the bin packing here is not to balance work in this job, it is to pack data into files so that work is balanced in future jobs.

I think it is probably a mistake to change this class.

File: core/src/main/java/org/apache/iceberg/util/BinPacking.java [hunk 7]
```diff
@@ -62,11 +62,6 @@ public ListPacker(long targetWeight, int lookback, boolean largestBinFirst) {
     private final Function<T, Long> weightFunc;
     private final boolean largestBinFirst;
 
-    public PackingIterable(Iterable<T> iterable, long targetWeight, int lookback,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/BinPacking.java:65]: Please revert this change. Even if this isn't used, the class is public and we don't need to make unnecessary breaking changes.

File: core/src/main/java/org/apache/iceberg/util/TableScanUtil.java [hunk 8]
```diff
@@ -56,7 +57,10 @@ public static boolean hasDeletes(FileScanTask task) {
     Preconditions.checkArgument(lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
     Preconditions.checkArgument(openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
 
-    Function<FileScanTask, Long> weightFunc = file -> Math.max(file.length(), openFileCost);
+    // For V2Format, we should check the size of delete file as well to avoid unbalanced bin-packing
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:60]: No need to use personal pronouns. This should be simply "check the size ... to avoid ..."

File: core/src/test/java/org/apache/iceberg/actions/TestBinPackStrategy.java [hunk 9]
```diff
@@ -69,6 +73,26 @@ public Table table() {
     return Arrays.stream(sizes).mapToObj(size -> new MockFileScanTask(size * MB)).collect(Collectors.toList());
   }
 
+  private FileScanTask fileOfDataAndDeleteSize(long dataSize, long[] deleteSizes) {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/actions/TestBinPackStrategy.java:76]: Nit: I think this should be `taskWithDataAndDeleteSizes`

File: core/src/test/java/org/apache/iceberg/actions/TestBinPackStrategy.java [hunk 10]
```diff
@@ -69,6 +73,26 @@ public Table table() {
...
+    DataFile dataFile = mockDataFile(dataSize);
+    DeleteFile[] deleteFiles = deleteFilesOfSize(deleteSizes);
+    return new MockFileScanTask(dataFile, deleteFiles);
+  }
+
+  private DeleteFile[] deleteFilesOfSize(long... sizes) {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/actions/TestBinPackStrategy.java:82]: What about `deleteFilesWithSizes`?

File: core/src/test/java/org/apache/iceberg/actions/TestBinPackStrategy.java [hunk 11]
```diff
@@ -69,6 +73,26 @@ public Table table() {
...
+    return Arrays.stream(sizes).mapToObj(size -> {
+      DeleteFile mockDeleteFile = Mockito.mock(DeleteFile.class);
+      Mockito.when(mockDeleteFile.fileSizeInBytes()).thenReturn(size * MB);
+      return mockDeleteFile;
+    }).toArray(DeleteFile[]::new);
+  }
+
+  private DataFile mockDataFile(long size) {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/actions/TestBinPackStrategy.java:90]: For consistency, what about `dataFileWithSize`?

File: core/src/main/java/org/apache/iceberg/util/TableScanUtil.java [hunk 12]
```diff
@@ -56,7 +57,10 @@ public static boolean hasDeletes(FileScanTask task) {
...
+    Function<FileScanTask, Long> weightFunc = file -> Math.max(
+        file.length() + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
+        (1 + file.deletes().size()) * openFileCost);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:63]: Looks like there is no test for this change.

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 6 above)
Comment (WinkerDu) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:226]: I think the `file groups` here depends the rewrite job scale, it should take into account how much delete files cost in each rewrite job, otherwise there could be 'job-level' unbalanced issue.
@rdblue @RussellSpitzer what do you think?

File: core/src/main/java/org/apache/iceberg/util/TableScanUtil.java (hunk 12 above)
Comment (WinkerDu) [reply core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:63]: A new commit with related UT has pushed, please check @rdblue

**Review by WinkerDu — COMMENTED** [review]
@rdblue a new commit with minor revision according your CR has pushed, please have a further review, thx

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 6 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:226]: I don't think this should be based on deletes.

File: api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java [hunk 13]
```diff
@@ -76,6 +76,12 @@
    */
   String TARGET_FILE_SIZE_BYTES = "target-file-size-bytes";
 
+  /**
+   * The estimated cost to open a file, used as a minimum weight when combining splits. By default this
+   * will use the "read.split.open-file-cost" value in the table properties of the table being updated.
+   */
+  String OPEN_FILE_COST = "open-file-cost";
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:83]: The other properties are `file-open-cost`, not `open-file-cost`.

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 6 above)
Comment (WinkerDu) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java]: @rdblue OK, I revert the code change of `BinPackStrategy`, just introduce deletes based `weightFunc` in `TableScanUtil`, please have a further review, thank you

File: core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  @Test
+  public void testPlanTaskWithDeleteFiles() {
+    List<FileScanTask> testFiles = tasksWithDataAndDeleteSizes(
+        Arrays.asList(
+            Pair.of(100L, new Long[] {20L, 20L}),
+            Pair.of(250L, new Long[0]),
+            Pair.of(100L, new Long[] {40L, 40L, 40L}),
+            Pair.of(510L, new Long[] {1L, 1L}),
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java:70]: I don't think that these weights test what you want. Won't the results be the same either way since this weighs 510 without delete files and the next file weighs 4 without delete files? If you want to the new feature, I'd set this to 500 and add 3 delete files to the next file scan task. Then it weighs 16 with delete files and 4 without so you know that delete files are taken into account.

File: core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    }).toArray(DeleteFile[]::new);
+  }
+
+  @Test
+  public void testPlanTaskWithDeleteFiles() {
+    List<FileScanTask> testFiles = tasksWithDataAndDeleteSizes(
+        Arrays.asList(
+            Pair.of(100L, new Long[] {20L, 20L}),
+            Pair.of(250L, new Long[0]),
+            Pair.of(100L, new Long[] {40L, 40L, 40L}),
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java:69]: I think you want at least one file scan task that weighs over 512 with delete files alone. One with 250 data and 3 delete files of 100 each would do it. Then that shouldn't get combined with anything if weighed properly.

File: core/src/main/java/org/apache/iceberg/util/TableScanUtil.java [hunk 16]
```diff
@@ -56,7 +57,10 @@ public static boolean hasDeletes(FileScanTask task) {
     Preconditions.checkArgument(lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
     Preconditions.checkArgument(openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
 
-    Function<FileScanTask, Long> weightFunc = file -> Math.max(file.length(), openFileCost);
+    // Check the size of delete file as well to avoid unbalanced bin-packing
+    Function<FileScanTask, Long> weightFunc = file -> Math.max(
+        file.length() + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:62]: Should also consider the cost when using different join algorithm between delete files and data files ? 

For the equality files, the cost to join data file is :  `file.recordCount() * sum(eqDeleteFile.recordCount()) *  avgRecordByteSize)` .

For the positional delete files, the cost to join data file is: `file.length() + sum(posFiles.fileSizeInBytes())`.

The current approach sounds like we are treating all the delete files are positional delete files...

Comment (WinkerDu) [reply core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:62]: Thanks for your reply @openinx . Actually I don't think it's necessary to separate delete file to eq-delete and pos-delete when calculating cost. besides the algorithm here would be complicated, the top-level logic to apply delete files is 
`applyEqDeletes(applyPosDeletes(records))`
Generally I think the base logic above for eq-delete and pos-delete is the same.

File: core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java (hunk 15 above)
Comment (WinkerDu) [reply core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java]: @rdblue 
Sorry for late reply since I am busy these day.
I've change the unit test case as following with open cost 50, lookback 3 and target weight 300: 
```
group 1. data: 100, delete: 50, 100
group 2. data: 50, delete: 1, 50
group 3. data: 50, delete: 100
group 4. data: 1, delete: 1, 1
group 5. data: 75, delete 75
```
plan tasks after bin-packing would be `{group 1}, {group 2, group 3}, {group 4, group 5}`
1. group 1 total weight : 100 + 50 + 100 = 300
2. group 2 and group 3 total weight: `(50 + 1 + 50) -> (50 + 50 + 50)` + 50 + 100 = 300
3. group 3 and group 4 total weight: `(1 + 1 + 1) -> (50 + 50 + 50)` + 75 + 75 = 300

Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java]: It is easier to read if you don't have everything add up to exactly the target weight. That way the reader doesn't need to think about edge cases.

**Review by rdblue — APPROVED** [review]

File: core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java (hunk 15 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java]: This isn't a blocker, so I'll merge this when tests pass.

**Comment by rdblue** [comment]
Merged. Thanks for updating this, @WinkerDu!
